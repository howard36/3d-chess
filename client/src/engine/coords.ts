export type Coord = { x: number; y: number; z: number };

// UI constants for levels, files, and ranks
export const LEVELS = ['A', 'B', 'C', 'D', 'E']; // z
export const FILES = ['a', 'b', 'c', 'd', 'e']; // x
export const RANKS = ['1', '2', '3', '4', '5']; // y

/**
 * Converts a Coord to a string in ZXY format (e.g., 'Aa1').
 * LEVEL (A-E) + FILE (a-e) + RANK (1-5)
 */
export function toZXY(c: Coord): string {
  if (
    c.z < 0 ||
    c.z >= LEVELS.length ||
    c.x < 0 ||
    c.x >= FILES.length ||
    c.y < 0 ||
    c.y >= RANKS.length
  ) {
    throw new Error(`Invalid coordinate: ${JSON.stringify(c)}`);
  }
  return LEVELS[c.z] + FILES[c.x] + RANKS[c.y];
}

/**
 * Converts a ZXY string (e.g., 'Aa1') to a Coord.
 */
export function fromZXY(s: string): Coord {
  if (!/^([A-E])([a-e])([1-5])$/.test(s)) {
    throw new Error(`Invalid ZXY string: ${s}`);
  }
  const z = LEVELS.indexOf(s[0]);
  const x = FILES.indexOf(s[1]);
  const y = RANKS.indexOf(s[2]);
  if (z === -1 || x === -1 || y === -1) {
    throw new Error(`Invalid ZXY string: ${s}`);
  }
  return { x, y, z };
}
