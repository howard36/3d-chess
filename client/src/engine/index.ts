export { Board } from './board';
export type { Coord } from './coords';
export { LEVELS, FILES, RANKS, toZXY, fromZXY } from './coords';
export { PieceType } from './pieces';
export type { Piece } from './pieces';
export type { Move } from './board';
