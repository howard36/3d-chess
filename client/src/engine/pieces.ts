export enum PieceType {
  King = 'King',
  Queen = 'Queen',
  Rook = 'Rook',
  Bishop = 'Bishop',
  Knight = 'Knight',
  Unicorn = 'Unicorn',
  Pawn = 'Pawn',
}

export interface Piece {
  type: PieceType;
  color: 'white' | 'black';
}

// Movement vectors for each piece type (3D chess, §2.4)
// Each vector is [dz, dx, dy] (z=level, x=file, y=rank)

export const ROOK_VECTORS: ReadonlyArray<[number, number, number]> = [
  [1, 0, 0],
  [-1, 0, 0], // up/down
  [0, 1, 0],
  [0, -1, 0], // right/left
  [0, 0, 1],
  [0, 0, -1], // forward/back
];

export const BISHOP_VECTORS: ReadonlyArray<[number, number, number]> = [
  [1, 1, 0],
  [1, -1, 0],
  [-1, 1, 0],
  [-1, -1, 0],
  [1, 0, 1],
  [1, 0, -1],
  [-1, 0, 1],
  [-1, 0, -1],
  [0, 1, 1],
  [0, 1, -1],
  [0, -1, 1],
  [0, -1, -1],
];

export const UNICORN_VECTORS: ReadonlyArray<[number, number, number]> = [
  [1, 1, 1],
  [1, 1, -1],
  [1, -1, 1],
  [1, -1, -1],
  [-1, 1, 1],
  [-1, 1, -1],
  [-1, -1, 1],
  [-1, -1, -1],
];

export const QUEEN_VECTORS: ReadonlyArray<[number, number, number]> = [
  ...ROOK_VECTORS,
  ...BISHOP_VECTORS,
  ...UNICORN_VECTORS,
];

export const KING_VECTORS = QUEEN_VECTORS;

// Knight: all (±2,±1,0) and permutations, excluding (0,0,0)
export const KNIGHT_VECTORS: ReadonlyArray<[number, number, number]> = (() => {
  const result: [number, number, number][] = [];
  const deltas = [2, 1, 0];
  for (const dz of deltas) {
    for (const dx of deltas) {
      for (const dy of deltas) {
        if (
          Math.abs(dz) + Math.abs(dx) + Math.abs(dy) === 3 &&
          [dz, dx, dy].filter((v) => v !== 0).length === 2
        ) {
          for (const sz of [-1, 1]) {
            for (const sx of [-1, 1]) {
              for (const sy of [-1, 1]) {
                result.push([dz * sz, dx * sx, dy * sy]);
              }
            }
          }
        }
      }
    }
  }
  // Remove duplicates
  return Array.from(new Set(result.map((v) => v.toString()))).map(
    (s) => s.split(',').map(Number) as [number, number, number],
  );
})();

// Pawn movement is handled separately due to color and promotion rules
