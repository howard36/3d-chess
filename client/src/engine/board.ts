import {
  Piece,
  PieceType,
  ROOK_VECTORS,
  BISHOP_VECTORS,
  UNICORN_VECTORS,
  QUEEN_VECTORS,
  KING_VECTORS,
  KNIGHT_VECTORS,
} from './pieces';
import { Coord, LEVELS, FILES, RANKS } from './coords';

export class Board {
  grid: (Piece | null)[][][];

  constructor() {
    // 5x5x5 grid, all null by default
    this.grid = Array.from({ length: LEVELS.length }, () =>
      Array.from({ length: FILES.length }, () =>
        Array.from({ length: RANKS.length }, () => null as Piece | null),
      ),
    );
  }

  isInside(coord: Coord): boolean {
    return (
      coord.z >= 0 &&
      coord.z < LEVELS.length &&
      coord.x >= 0 &&
      coord.x < FILES.length &&
      coord.y >= 0 &&
      coord.y < RANKS.length
    );
  }

  setPiece(coord: Coord, piece: Piece | null): void {
    this.grid[coord.z][coord.x][coord.y] = piece;
  }

  getPiece(coord: Coord): Piece | null {
    return this.grid[coord.z][coord.x][coord.y];
  }

  generateMoves(from: Coord): Coord[] {
    const piece = this.getPiece(from);
    if (!piece) return [];
    // Pawn logic
    if (piece.type === PieceType.Pawn) {
      const moves: Coord[] = [];
      const dir = piece.color === 'white' ? 1 : -1;
      // Forward (y axis)
      const forward: Coord = { x: from.x, y: from.y + dir, z: from.z };
      if (this.isInside(forward) && !this.getPiece(forward)) {
        moves.push(forward);
      }
      // Up (z axis)
      const up: Coord = { x: from.x, y: from.y, z: from.z + dir };
      if (this.isInside(up) && !this.getPiece(up)) {
        moves.push(up);
      }
      // Captures: 5 specified diagonal directions
      // Spec Section 2.5: Relative to White (dir=1):
      // - Forwards-Up: (0, +1, +1) -> (dx=0, dy=dir, dz=dir)
      // - Forwards-Left: (-1, +1, 0) -> (dx=-1, dy=dir, dz=0)
      // - Forwards-Right: (+1, +1, 0) -> (dx=1, dy=dir, dz=0)
      // - Up-Left: (-1, 0, +1) -> (dx=-1, dy=0, dz=dir)
      // - Up-Right: (+1, 0, +1) -> (dx=1, dy=0, dz=dir)
      const captureDeltas: [number, number, number][] = [
        [0, dir, dir], // Forwards-Up
        [-1, dir, 0], // Forwards-Left
        [1, dir, 0], // Forwards-Right
        [-1, 0, dir], // Up-Left
        [1, 0, dir], // Up-Right
      ];

      // Only allow captures to squares with enemy piece
      for (const [dx, dy, dz] of captureDeltas) {
        const to: Coord = { x: from.x + dx, y: from.y + dy, z: from.z + dz };
        if (this.isInside(to)) {
          const target = this.getPiece(to);
          if (target && target.color !== piece.color) {
            moves.push(to);
          }
        }
      }
      return moves;
    }
    let vectors: ReadonlyArray<[number, number, number]> = [];
    let sliding = false;
    switch (piece.type) {
      case PieceType.Rook:
        vectors = ROOK_VECTORS;
        sliding = true;
        break;
      case PieceType.Bishop:
        vectors = BISHOP_VECTORS;
        sliding = true;
        break;
      case PieceType.Unicorn:
        vectors = UNICORN_VECTORS;
        sliding = true;
        break;
      case PieceType.Queen:
        vectors = QUEEN_VECTORS;
        sliding = true;
        break;
      case PieceType.King:
        vectors = KING_VECTORS;
        sliding = false;
        break;
      case PieceType.Knight:
        vectors = KNIGHT_VECTORS;
        sliding = false;
        break;
      default:
        return [];
    }
    const moves: Coord[] = [];
    for (const [dz, dx, dy] of vectors) {
      let n = 1;
      while (true) {
        const to: Coord = { z: from.z + dz * n, x: from.x + dx * n, y: from.y + dy * n };
        if (!this.isInside(to)) break;
        const target = this.getPiece(to);
        if (!target) {
          moves.push(to);
        } else {
          if (target.color !== piece.color) moves.push(to);
          break;
        }
        if (!sliding) break;
        n++;
      }
    }
    return moves;
  }

  applyMove(from: Coord, to: Coord, promotion?: PieceType, board: Board = this): void {
    const piece = board.getPiece(from);
    if (!piece) return;
    // Remove from origin
    board.setPiece(from, null);
    // Promotion logic
    let newPiece = piece;
    if (
      piece.type === PieceType.Pawn &&
      ((piece.color === 'white' && to.y === 4 && to.z === 4) ||
        (piece.color === 'black' && to.y === 0 && to.z === 0)) &&
      promotion &&
      promotion !== PieceType.Pawn &&
      promotion !== PieceType.King
    ) {
      newPiece = { type: promotion, color: piece.color };
    }
    board.setPiece(to, newPiece);
  }

  findKing(color: 'white' | 'black'): Coord {
    for (let z = 0; z < LEVELS.length; z++) {
      for (let x = 0; x < FILES.length; x++) {
        for (let y = 0; y < RANKS.length; y++) {
          const piece = this.getPiece({ x, y, z });
          if (piece && piece.type === PieceType.King && piece.color === color) {
            return { x, y, z };
          }
        }
      }
    }
    throw new Error(`King of color ${color} not found`);
  }

  isSquareAttacked(target: Coord, byColor: 'white' | 'black'): boolean {
    for (let z = 0; z < LEVELS.length; z++) {
      for (let x = 0; x < FILES.length; x++) {
        for (let y = 0; y < RANKS.length; y++) {
          const piece = this.getPiece({ x, y, z });
          if (piece && piece.color === byColor) {
            const from = { x, y, z };
            const moves = this.generateMoves(from);
            if (moves.some((m) => m.x === target.x && m.y === target.y && m.z === target.z)) {
              return true;
            }
          }
        }
      }
    }
    return false;
  }

  /**
   * Returns true if the king of the given color is in check.
   */
  inCheck(color: 'white' | 'black'): boolean {
    const kingPos = this.findKing(color);
    const enemyColor = color === 'white' ? 'black' : 'white';
    return this.isSquareAttacked(kingPos, enemyColor);
  }

  /**
   * Generate all legal moves for the given color.
   * Returns array of { from, to, promotion? }
   */
  generateLegalMoves(
    color: 'white' | 'black',
  ): { from: Coord; to: Coord; promotion?: PieceType }[] {
    const legalMoves: { from: Coord; to: Coord; promotion?: PieceType }[] = [];
    for (let z = 0; z < LEVELS.length; z++) {
      for (let x = 0; x < FILES.length; x++) {
        for (let y = 0; y < RANKS.length; y++) {
          const piece = this.getPiece({ x, y, z });
          if (piece && piece.color === color) {
            const from: Coord = { x, y, z };
            const moves = this.generateMoves(from);
            for (const to of moves) {
              // Handle pawn promotion: if pawn moves to promotion square, generate all valid promotions
              if (
                piece.type === PieceType.Pawn &&
                ((piece.color === 'white' && to.y === 4 && to.z === 4) ||
                  (piece.color === 'black' && to.y === 0 && to.z === 0))
              ) {
                // Try all promotion types except Pawn/King
                for (const promotion of [
                  PieceType.Queen,
                  PieceType.Rook,
                  PieceType.Bishop,
                  PieceType.Knight,
                  PieceType.Unicorn,
                ]) {
                  const clone = this.clone();
                  clone.applyMove(from, to, promotion);
                  if (!clone.inCheck(color)) {
                    legalMoves.push({ from, to, promotion });
                  }
                }
              } else {
                const clone = this.clone();
                clone.applyMove(from, to);
                if (!clone.inCheck(color)) {
                  legalMoves.push({ from, to });
                }
              }
            }
          }
        }
      }
    }
    return legalMoves;
  }

  /**
   * Returns true if the given color is checkmated.
   */
  isCheckmate(color: 'white' | 'black'): boolean {
    return this.inCheck(color) && this.generateLegalMoves(color).length === 0;
  }

  /**
   * Returns true if the given color is stalemated.
   */
  isStalemate(color: 'white' | 'black'): boolean {
    return !this.inCheck(color) && this.generateLegalMoves(color).length === 0;
  }

  /**
   * Returns true if the given color has at least one legal move.
   */
  hasLegalMove(color: 'white' | 'black'): boolean {
    return this.generateLegalMoves(color).length > 0;
  }

  /**
   * Shallow clone of the board (for move simulation)
   */
  clone(): Board {
    const newBoard = new Board();
    // Deep copy grid (3D array)
    newBoard.grid = this.grid.map((level) => level.map((file) => file.slice()));
    return newBoard;
  }
}
