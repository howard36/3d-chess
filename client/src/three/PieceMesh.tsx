import React from 'react';
import type { JSX } from 'react';
import { PieceType } from '../engine';

export type PieceMeshProps = JSX.IntrinsicElements['mesh'] & {
  type: PieceType;
  color: 'white' | 'black';
};

const pieceColor = (color: 'white' | 'black') => (color === 'white' ? 0xffffff : 0x222222);

export const PieceMesh: React.FC<PieceMeshProps> = ({ type, color, ...props }) => {
  // Each piece gets a different primitive mesh
  switch (type) {
    case PieceType.Pawn:
      return (
        <mesh {...props}>
          <cylinderGeometry args={[0.25, 0.25, 0.6, 16]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    case PieceType.Rook:
      return (
        <mesh {...props}>
          <cylinderGeometry args={[0.3, 0.3, 0.7, 8]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    case PieceType.Bishop:
      return (
        <mesh {...props}>
          <coneGeometry args={[0.28, 0.8, 12]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    case PieceType.Knight:
      return (
        <mesh {...props}>
          <cylinderGeometry args={[0.2, 0.2, 0.5, 6]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    case PieceType.Unicorn:
      return (
        <mesh {...props}>
          <coneGeometry args={[0.18, 0.7, 6]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    case PieceType.Queen:
      return (
        <mesh {...props}>
          <cylinderGeometry args={[0.22, 0.32, 0.9, 16]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    case PieceType.King:
      return (
        <mesh {...props}>
          <cylinderGeometry args={[0.25, 0.35, 1.0, 16]} />
          <meshStandardMaterial color={pieceColor(color)} />
        </mesh>
      );
    default:
      return null;
  }
};
