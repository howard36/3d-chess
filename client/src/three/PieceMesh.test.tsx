import React from 'react';
import { render } from '@testing-library/react';
import { PieceMesh } from './PieceMesh';
import { PieceType } from '../engine';
import '@testing-library/jest-dom';

const types = [
  PieceType.Pawn,
  PieceType.Rook,
  PieceType.Bishop,
  PieceType.Knight,
  PieceType.Unicorn,
  PieceType.Queen,
  PieceType.King,
];

describe('PieceMesh', () => {
  types.forEach((type) => {
    ['white', 'black'].forEach((color) => {
      it(`renders ${color} ${type} and matches snapshot`, () => {
        const { container } = render(<PieceMesh type={type} color={color as 'white' | 'black'} />);
        expect(container).toMatchSnapshot();
      });
    });
  });
});
