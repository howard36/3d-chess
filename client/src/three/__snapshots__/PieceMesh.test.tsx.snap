// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PieceMesh > renders black Bishop and matches snapshot 1`] = `
<div>
  <mesh>
    <conegeometry
      args="0.28,0.8,12"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders black King and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.25,0.35,1,16"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders black Knight and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.2,0.2,0.5,6"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders black Pawn and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.25,0.25,0.6,16"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders black Queen and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.22,0.32,0.9,16"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders black Rook and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.3,0.3,0.7,8"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders black Unicorn and matches snapshot 1`] = `
<div>
  <mesh>
    <conegeometry
      args="0.18,0.7,6"
    />
    <meshstandardmaterial
      color="2236962"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white Bishop and matches snapshot 1`] = `
<div>
  <mesh>
    <conegeometry
      args="0.28,0.8,12"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white King and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.25,0.35,1,16"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white Knight and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.2,0.2,0.5,6"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white Pawn and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.25,0.25,0.6,16"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white Queen and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.22,0.32,0.9,16"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white Rook and matches snapshot 1`] = `
<div>
  <mesh>
    <cylindergeometry
      args="0.3,0.3,0.7,8"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;

exports[`PieceMesh > renders white Unicorn and matches snapshot 1`] = `
<div>
  <mesh>
    <conegeometry
      args="0.18,0.7,6"
    />
    <meshstandardmaterial
      color="16777215"
    />
  </mesh>
</div>
`;
